### PR TITLE
use file name instead of checksum to check whether a migration was run

### PR DESCRIPTION
**Pull request by olivierdalang**
As per https://github.com/opengisch/pum/issues/79

This is a draft, not sure of the approach... would probably be worth keeping checksum verification to display warnings.

### CODE REVIEW

**Comment by 3nids** [comment]
it makes sense, no objection on my side

**Comment by olivierdalang** [comment]
The test failed because in the test suite, there are cases where two migrations have the same file name, but are in a different folders.

With the old logic, both run, because they had different checksums anyways.
With the new logic, only the first runs, as the second time, it's considered already applied.

Is there a proper use case for two identically named files in different dirs representing different migrations ? Is it used somewhere ?
This could not be supported anymore as there no way to properly distinguish the folders (sounds like a bad idea anyways :-) ) .

So, with this PR, the migrations are identified using `version`, `description`, `type`, where any two files producing the same values would be considered run already.

**Comment by 3nids** [comment]
I believe this is fine.
This might be worth adding some notes in the README.
Shall I merge?

**Comment by olivierdalang** [comment]
@3nids 
> Shall I merge?

From a strictly-QGEP perspective, yes, but I'm not really aware of usage outside of QGEP.

**Comment by haubourg** [comment]
> From a strictly-QGEP perspective, yes, but I'm not really aware of usage outside of QGEP.

I'd ask @elemoine here (hi Eric!)

**Comment by elemoine** [comment]
Thanks for notifying me @haubourg!

No strong opinion here.

One thing though: this [docstring](https://github.com/opengisch/pum/blob/a6655cc158f4b13693987a4b48e4222a8cae1775/pum/core/upgrader.py#L507-L523) in the code suggests to use a date in the form `ddmmyy` as the delta name. And there may be other places where the same suggestion is made. If `__is_applied` is changed to rely on (version, name, type) rather than (version, checksum) we certainly want to suggest giving unique names to deltas.

**Comment by ponceta** [comment]
@olivierdalang status on this? This was a good or a bad idea considering your experiences?

**Comment by olivierdalang** [comment]
I think the change is good for the reason described in the initial ticket.

From a QGEP perspective, we can merge, because the case of two different migrations with the same name does not happen in QGEP. Maybe if we do that, we should completely remove the ability to specify multiple delta directories (for which I don't know the use case anyway), so that name clashes in migrations are not possible anymore.

(however I can't tell if it's worth putting effort in this currently)

File: pum/core/upgrader.py [hunk 1]
```diff
@@ -351,14 +351,14 @@ def __is_applied(self, delta):
...
         query = """
         SELECT id FROM {}
         WHERE version = '{}'
-            AND checksum = '{}'
+            AND description = '{}'
+            AND type = '{}'
             AND success = 'TRUE'
         """.format(
-            self.upgrades_table, delta.get_version(), delta.get_checksum())
+            self.upgrades_table, delta.get_version(), delta.get_name(), delta.get_type())
```
Comment (3nids) [inline pum/core/upgrader.py:361]: if the other PR gets through, let's do formatted strings here too :)

**Review by 3nids — APPROVED** [review]

**Comment by olivierdalang** [comment]
Still slightly scared, but we got to merge it at some point... As far as I understood, the only know case of multi delta dirs is for [sigip](https://github.com/qwat/qwat-data-model/tree/master/.build/customizations/sigip/delta), where there's just a custom pre and post-all.

So merging and let's hope for the best :-)

**Pull request merged**